oml/functional/metrics.py

```diff
@@ -465,15 +486,73 @@ def calc_fnmr_at_fmr(pos_dist: torch.Tensor, neg_dist: torch.Tensor, fmr_vals: T
         tensor([40., 20.])
 
     """
-    if len(fmr_vals) == 0:
-        raise ValueError(f"fmr_vals are expected have at least one value, but got {fmr_vals}")
-    if not all(0 <= f <= 100 for f in fmr_vals):
-        raise ValueError(f"fmr_vals are expected to be integers in range [0, 100] but got {fmr_vals}")
+    _check_if_in_range(fmr_vals, 0, 100, "fmr_vals")
     thresholds = torch.from_numpy(np.percentile(neg_dist.cpu().numpy(), fmr_vals)).to(pos_dist)
     fnmr_at_fmr = 100 * (pos_dist[None, :] >= thresholds[:, None]).sum(axis=1) / len(pos_dist)
     return fnmr_at_fmr
 
 
+def calc_main_components_percentage(
+    embeddings: torch.Tensor, explained_variance_to_keep: Tuple[int, ...]
+) -> List[torch.Tensor]:
+    """
+    Function estimates the number of main components of embeddings using Principal Component Analysis.
+    The number of components is estimated as a percentage of the embeddings dimension.
+
+    Args:
+        embeddings: Embeddings matrix with the shape of ``[n_embeddings, embeddings_dim]``.
+        explained_variance_to_keep: Values in range [0, 100]. Find the number of components such that the amount
+                                    of variance that needs to be explained is greater than the percentage specified
+                                    by ``explained_variance_to_keep``.
+    Returns:
+        List of linear dimensions  as percentages of the embeddings dimension.
+
+    Let :math:`X` be a set of :math:`d` dimensional embeddings.
+    Let :math:`\\lambda_1, \\ldots, \\lambda_d\\in\\mathbb{R}` be a set of eigenvalues
+    of the covariance matrix of :math:`X` sorted in descending order.
+    Then for a given value of desired explained variance :math:`r`,
+    the number of main components that explaines :math:`r\\cdot 100\\%%` variance is the largest integer
+    :math:`n` such that
+
+    .. math::
+        \\frac{\\sum\\limits_{i = 1}^{n - 1}\\lambda_i}{\\sum\\limits_{i = 1}^{d}\\lambda_i} \\leq r
+
+    The function returns
+
+    .. math::
+        \\frac{n}{d}\\cdot 100\\%%
+
+    See:
+
+        `Principal Components Analysis`_
+
+    .. _`Principal Components Analysis`:
+        https://en.wikipedia.org/wiki/Principal_component_analysis
+
+    Example:
+        >>> embeddings = torch.tensor([
+        ...                            [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+        ...                            [0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
+        ...                            [0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
+        ... ], dtype=torch.float)
+        >>> calc_main_components_percentage(embeddings, explained_variance_to_keep=(50, 100))
+        tensor([10., 40.])
+
+    """
+    # The code below mirrors code from scikit-learn repository:
+    # https://github.com/scikit-learn/scikit-learn/blob/f3f51f9b6/sklearn/decomposition/_pca.py#L491
```

Comment:
It looks like this should be not a function but a class. If it will be a function, it will return, for instance, `mean`, `singular_values`, `components` etc. But in future, to apply all this intermediate results for new vectors one has to know what should be done with all these values...

So I'd suggest to implement a simple class `PCA` like this

<details>
<summary><tt>PCA</tt></summary>

```python

class PCA:
    components: torch.Tensor
    mean: torch.Tensor
    singular_values: torch.Tensor
    explained_variance: torch.Tensor
    explained_variance_ratio: torch.Tensor

    def __init__(self, embeddings: torch.Tensor):
        self._fit(embeddings)
    
    def _fit(self, embeddings: torch.Tensor):
        ...
        self.mean = embeddings.mean(dim=0).unsqueeze(1)
        self.components = ...
        ...

    def transform(self, embeddings: torch.Tensor):
        ...
        embeddings = embeddings - self.mean
        ...
```
</details>


Afterwards all the `mean`, `explained_variance` etc. could be used as it is required to achieve some goals. For instance in my case I will be using only `explained_variance_ratio`.

Reply:
@dapladoc class is even better